Update the generic implementation model (#1398)
carbon-language/carbon-lang: docs/design/generics/details.md

[deduced parameters](overview.md#deduced-parameters) in the Generics overview
doc). Basically, the user passes in a value for `val`, and the type of `val`
determines `T`. `T` still gets passed into the function though, and it plays an
important role -- it defines the key used to look up interface implementations.

We can think of the interface as defining a struct type whose members are
function pointers, and an implementation of an interface as a value of that
struct with actual function pointer values. So an implementation is a table of

> function pointers, one per function defined in the interface. For more on this,

Reviewer: ```suggestion
struct with actual function pointer values. An implementation is a table mapping
the interface's function signatures to function pointers. For more on this,
```

I think this sentence is currently hard to understand:

- One table per function, or one pointer per function? I assume pointer per function.
- If it's just a table of function pointers, how does that list work? I assume it's a mapping.

Trying to suggest something that I think may be clearer.

Author: I replace "function signatures" with "functions"

-   [Interfaces](#interfaces) are types of witness tables.
-   [Impls](#implementing-interfaces) are witness table values.

Type checking is done with just the interface. The impl is used during code
generation time, possibly using
[monomorphization](https://en.wikipedia.org/wiki/Monomorphization) to have a
separate instantiation of the function for each combination of the generic
argument values. The compiler is free to use other implementation strategies,

> though, such as passing the witness table for any needed implementations, if

Reviewer: ```suggestion
such as passing the witness table for any needed implementations, if
```

};
```

> Since generic arguments, where the parameter is declared using `:!`, are passed

Reviewer: ```suggestion
Since generic arguments (where the parameter is declared using `:!`) are passed
```

Suggest swapping to parens to make comma-related reading of this easier.

at compile time, so the actual value of `VectorForPoint` can be used to generate
the code for functions using that impl. This is the

> [static-dispatch witness table](terminology.md#static-dispatch-witness-table).

Reviewer: ```suggestion
[static-dispatch witness table](terminology.md#static-dispatch-witness-table)
```
